font/src/ft/mod.rs

```diff
@@ -213,75 +222,56 @@ impl FreeTypeRasterizer {
-                self.get_specific_face(&desc, &style)
+                // If a name was specified, try and load specifically that font
+                pattern.add_style(style);
             },
         }
-    }
-
-    fn full_metrics(&self, key: FontKey) -> Result<FullMetrics, Error> {
-        let face = self.faces.get(&key).ok_or(Error::FontNotLoaded)?;
-
-        let size_metrics = face.ft_face.size_metrics().ok_or(Error::MissingSizeMetrics)?;
-
-        let width = match face.ft_face.load_char('0' as usize, face.load_flags) {
-            Ok(_) => face.ft_face.glyph().metrics().horiAdvance / 64,
-            Err(_) => size_metrics.max_advance / 64,
-        } as f64;
-
-        Ok(FullMetrics { size_metrics, cell_width: width })
-    }
-
-    fn get_matching_face(
-        &mut self,
-        desc: &FontDesc,
-        slant: Slant,
-        weight: Weight,
-    ) -> Result<FontKey, Error> {
-        let mut pattern = Pattern::new();
-        pattern.add_family(&desc.name);
-        pattern.set_weight(weight.into_fontconfig_type());
-        pattern.set_slant(slant.into_fontconfig_type());
-        pattern.add_pixelsize(self.pixel_size);
 
-        self.query_font(pattern, desc)
-    }
-
-    fn get_specific_face(&mut self, desc: &FontDesc, style: &str) -> Result<FontKey, Error> {
-        let mut pattern = Pattern::new();
-        pattern.add_family(&desc.name);
-        pattern.add_style(style);
-        pattern.add_pixelsize(self.pixel_size);
+        // Match pattern and get a font list, use first font as a "primary" one and the rest for a
+        // font fallback
+        let matched_fonts = fc::font_sort(&config, &mut pattern.clone())
+            .ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+        let mut matched_fonts = matched_fonts.into_iter();
 
-        self.query_font(pattern, desc)
-    }
+        let primary_font =
+            matched_fonts.next().ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
 
-    fn query_font(&mut self, pattern: Pattern, desc: &FontDesc) -> Result<FontKey, Error> {
-        let config = fc::Config::get_current();
-        let fonts = fc::font_sort(&config, &mut pattern.clone())
-            .ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+        // We should render patterns to get values like `pixelsizefixupfactor`
+        let primary_font = pattern.render_prepare(config, primary_font);
 
-        let mut font_iter = fonts.into_iter();
+        // Compute ID for font identification, we're using hashes from requested pattern
+        // and a rendered one, and combining both with XOR to get a resulted ID. Some fonts
+        // always have the same size after pattern matching match, so if user requests pattern
+        // with size `X`, resulted font could always have `size = const`, which leads to
+        // constant result from e.g. `primary_font.hash()`, so we need to use requested
+        // `pixelsize` in the hash computation in addition to the hash of rendered pattern
```

Comment:
I think I've asked you this before, but what happens if we set the font size for a font which loses the size during pattern prepare?

If we set it for one that doesn't lose it, it shouldn't make any difference, right?

What happens if we manually set it again after preparing it for all fonts?

Reply:
>I think I've asked you this before, but what happens if we set the font size for a font which loses the size during pattern prepare?

size is always with *us*, you can't just lose it. The thing is that some fonts have constant size. So if you request a size 12 for font "Emoji" which has fixed size 108, you'll get size 108 and a fixupfactor to scale thing. For scaled fonts you're getting a size more realistic to what you request and it based on your request of 12.

>If we set it for one that doesn't lose it, it shouldn't make any difference, right?

Again, you're not losing anything, you're just always getting a constant size no matter what you try to request from it.

>What happens if we manually set it again after preparing it for all fonts?

If you manually set size to a already rendered pattern, it'll blow up.

Reply:
> If you manually set size to a already rendered pattern, it'll blow up.

Huh, interesting. Mind telling me how it blows up? Will it just segfault?

Reply:
No, it messes up your "computed" fields, etc. So you'll get "EXTRA BIG EMOJIS" for example, since you're modifying the returned `size` for your previous request. Like if you force pixelsize of 12 to a thing with constant size 108, your fixup factor would be based off your forced value, however underlining bitmap metrics are completely different (they match this constant size).